compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp

```diff
@@ -186,30 +164,237 @@ static bool isEligibleForCollapse(linalg::GenericOp genericOp) {
+//===---------------------------------------------------------------------===//
+
+namespace {
+class CollapseInfo {
+public:
+  using CollapsableLoopsSet = llvm::SmallSetVector<int64_t, 8>;
+
+  CollapseInfo() = default;
+  CollapseInfo(ArrayRef<ReassociationIndices> reassociation)
+      : reassociation(reassociation),
+        collapsableLoops(getCollapsedFromReassociation(reassociation)) {}
+
+  // Print the current operation & reassociation indicies
+  void print(raw_ostream &os) const;
+
+  // Debug print the current operation & reassociation indicies
+  void dump() const;
+
+  // Update `collapsableLoops` by taking the set intersection with
+  // `otherCollapsable` and update the reassociation indicies accordingly.
+  void updateCollapseViaIntersect(const CollapsableLoopsSet &otherCollapsable);
+
+  // Update `collapsableLoops` by subtracting `uncollapsable` and update the
+  // reassociation indicies accordingly.
+  void updateCollapseViaSubtract(const CollapsableLoopsSet &uncollapsable);
+
+  // Get `collapsableLoops` after applying the transformation provided by `map`.
+  // Note: doesn't modify `collapsableLoops`, the tranformation is applied to a
+  // copy.
+  FailureOr<CollapsableLoopsSet>
+  getTransformedCollapsableLoops(AffineMap map) const;
+
+  // Clear internal data
+  void clear() {
+    reassociation.clear();
+    collapsableLoops.clear();
+  }
+
+  const CollapsableLoopsSet &getCollapsibleLoops() const {
+    return collapsableLoops;
+  }
+
+  const SmallVector<ReassociationIndices> &getReassocation() const {
+    return reassociation;
+  }
+
+private:
+  static CollapsableLoopsSet
+  getCollapsedFromReassociation(ArrayRef<ReassociationIndices> reassociation) {
+    CollapsableLoopsSet collapsed;
+    for (auto &indicies : reassociation) {
+      for (int64_t index : indicies) {
+        collapsed.insert(index);
+      }
+    }
+    return collapsed;
+  }
+
+  void updateReassociation();
+
+private:
+  SmallVector<ReassociationIndices> reassociation;
+
+  // Note: `collapsableLoops` does not directly map to `reassociation`
+  // because parallel and reduction iteration dimensions must be kept separate.
+  CollapsableLoopsSet collapsableLoops;
+};
+} // namespace
+
+void CollapseInfo::updateReassociation() {
+  SmallVector<ReassociationIndices> newReassociation;
+  for (auto &indicies : reassociation) {
+    ReassociationIndices newIndicies;
+    for (int64_t index : indicies) {
+      if (collapsableLoops.contains(index)) {
+        newIndicies.push_back(index);
+        continue;
+      }
+
+      // Current index is not collapsable
+      if (newIndicies.size() > 1) {
+        newReassociation.push_back(newIndicies);
+      }
+      newIndicies.clear();
+    }
+
+    if (newIndicies.size() > 1) {
+      newReassociation.push_back(newIndicies);
+    }
+  }
+  reassociation = std::move(newReassociation);
+}
+
+FailureOr<CollapseInfo::CollapsableLoopsSet>
+CollapseInfo::getTransformedCollapsableLoops(AffineMap map) const {
```

Comment:
Here too... write an example of what the transformedloops would be for a given map and collapsableLoops.